akka-actor/src/main/scala/akka/actor/dungeon/TimerSchedulerImpl.scala

```diff
@@ -45,7 +45,10 @@ import akka.util.OptionVal
     }
     val nextGen = nextTimerGen()
 
-    val timerMsg = TimerMsg(key, nextGen, this)
+    val timerMsg = msg match {
+      case _: AutoReceivedMessage ⇒ msg
```

Comment:
This will not work. Then all checks that are done when receiving a TimerMsg is not done. I think nothing should be done here and instead in `Timers.aroundReceive` add something like:

```
          case OptionVal.Some(m: AutoReceivedMessage) ⇒
            self.asInstanceOf[LocalActorRef].underlying.autoReceiveMessage(Envelope(m, sender()))
```

It would be good to add a test also.

Reply:
Your approach works for Actor's that composed with `Timers` trait.

Also same bug appears in `akka.actor.FSM` class.  `akka.actor.FSM.setTimer` method creates a `akka.actor.FSM.Timer` and calls `schedule` method and it wraps `AutoReceivedMessage` with `akka.actor.FSM.Timer`. So FSM Actor actually receives `PoisonPill` or any other AutoReceivedMessage.

Possible solution is again using  in `akka.actor.FSM.Timer.schedule` method, something like:

``` 
val timerMsg = msg match {
        case m: AutoReceivedMessage ⇒ m
        case _                      ⇒ this
      }
``` 
Is it convenient for you?

Reply:
@mucahitkantepe yes, it's fine to do similar fix in FSM

Reply:
and `PersistentFSM` :(